File: 0091-decode-ways/0091-decode-ways.java
class Solution {
    
    Map<Integer, Integer> memo = new HashMap<>();
    
    public int numDecodings(String s) {
        return numDecoding(s, 0);
    }
    
    private int numDecoding(String s, int start) {
        
        if (memo.containsKey(start)) {
            return memo.get(start);
        }
        
        if (start == s.length()) {
            return 1;
        }
        
        if (s.charAt(start) == '0') {
                return 0;
        }
        
        if (start == s.length() - 1) {
            return 1;
        }
        
        int count = numDecoding(s, start + 1);
        if (start + 1 < s.length()) {
            Integer val = Integer.valueOf(s.substring(start, start + 2));
            if (val <= 26) {
                count += numDecoding(s, start + 2);
            }
        }
        
        memo.put(start, count);
        
        return count;
    }
}

File: 0091-decode-ways/0091-decode-ways.py
class Solution:
    
    def numDecodings(self, s: str) -> int:
        memo = {}
        return self.numDecoding(s, 0, memo)
    
    def numDecoding(self, s: str, index:int, memo) -> int:
        if memo.get(index):
            return memo.get(index)
        
        if index == len(s):
            return 1
        

        if s[index] == "0":
            return 0

        
        
        if index + 1 == len(s):
            return 1
    
        count = self.numDecoding(s, index + 1, memo)
        
        if index + 1 < len(s): 
            i = int(s[index :index + 2])
            
            if i <= 26:
                count += self.numDecoding(s, index + 2, memo)
        
        memo[index] = count
        return count